File: ports-rb/ports.rb
require "json"
require "test/unit"
require_relative "scheme"

class PortsFunction
  attr_accessor :function, :env

  def initialize(func, env)
    @function = func
    @env = env
  end

  def valid?
    !@function.nil?
  end

  def call(*args)
    @function.call(@env, *args)
  end

  def [](key)
    if key == 0
      ports_role
    elsif key == 1
      self
    end
  end
end

class PortsSetup < PortsFunction
  def ports_role
    :setup
  end
end

class PortsTearDown < PortsFunction
  def ports_role
    :tearDown
  end
end

class Placeholder < PortsFunction
  attr_accessor :name, :parameters, :doc_string

  def initialize(name, parameters, doc_string)
    super(nil, nil)
    @name = name
    @parameters = parameters
    @doc_string = doc_string
  end

  def [](key)
    :placeholder if key == 0
  end

  def template_string
    params = ["env"]
    params += parameters.map { |param| param.to_s.tr("-", "_") }

    <<~TEMPLATE
      placeholder "#{name}" do |#{params.join(", ")}|
        # TODO: Implement
      end
    TEMPLATE
  end
end

def ports_assert(value, msg = "")
  raise Test::Unit::AssertionFailedError, msg || "Expected #{value} to be truthy" unless value
end

def ports_assert_eq(expected, actual)
  raise Test::Unit::AssertionFailedError, "Expected #{actual} to be #{expected}" unless expected == actual
end

class PortsSuite
  attr_reader :scheme_env, :suite_name, :suite_version, :sources,
    :placeholders, :root_capability

  def initialize(file_name)
    @scheme_env = Scheme::Environment.new([], [], Scheme::GLOBAL_ENV)
    initialize_ports_primitives
    initialize_ports
    @suite_source = File.read(file_name)

    @suite_name = nil
    @suite_version = nil
    @sources = nil
    @placeholders = nil
    @root_capability = nil

    @placeholder_functions = {}
    @set_up_functions = []
    @tear_down_functions = []
  end

  def initialize_suite
    @suite_name, @suite_version, @sources,
    @placeholders, @root_capability = eval_scheme(@suite_source)
  end

  def eval_scheme(code, env = nil)
    env ||= @scheme_env
    Scheme.evaluate_string(code, env)
  end

  def eval_scheme_with_args(code, **kwargs)
    env = Scheme::Environment.new([], [], @scheme_env)
    kwargs.each do |key, value|
      env[key.to_sym] = value
    end
    eval_scheme(code, env)
  end

  def create_placeholder(name, parameters, doc_string = "")
    new_placeholder = Placeholder.new(name, parameters, doc_string)
    @scheme_env[name] = new_placeholder
    new_placeholder.env = @scheme_env

    if @placeholder_functions[name]
      new_placeholder.function = @placeholder_functions[name]
    end

    new_placeholder
  end

  def initialize_ports_primitives
    primitives = {
      "create-placeholder" => lambda { |*args| create_placeholder(*args) },
      "is-placeholder?" => lambda { |x| x.is_a?(Placeholder) },
      "assert" => method(:ports_assert),
      "assert-equal" => method(:ports_assert_eq),
      "is-assertion-error?" => lambda { |e| e.is_a?(Test::Unit::AssertionFailedError) }
    }

    primitives.each do |key, value|
      @scheme_env[key.to_sym] = value
    end
  end

  def initialize_ports
    ports_content = File.read("ports/ports.scm")
    eval_scheme(ports_content)
  end

  def placeholder(name, &func)
    @placeholder_functions[name.to_sym] = func
    func
  end

  def set_up(&func)
    @set_up_functions << func
    func
  end

  def tear_down(&func)
    @tear_down_functions << func
    func
  end

  def ensure_placeholders_are_valid
    invalid_placeholders = @placeholders.reject(&:valid?)
    return if invalid_placeholders.empty?

    invalid_placeholder_list = invalid_placeholders.map { |p| "- #{p.name}" }.join("\n")
    placeholders_suggestion = invalid_placeholders.map(&:template_string).join("\n")
    raise [
      "Your test suite is missing definitions for the following placeholders:\n#{invalid_placeholder_list}\n",
      "Implement the missing placeholders:\n#{placeholders_suggestion}"
    ].join("\n")
  end

  def install_set_up_tear_down_functions
    @set_up_functions.each do |func|
      @root_capability[2].unshift(PortsSetup.new(func, @scheme_env))
    end
    @tear_down_functions.each do |func|
      @root_capability[3].unshift(PortsTearDown.new(func, @scheme_env))
    end
  end

  def run(config)
    initialize_suite
    install_set_up_tear_down_functions
    ensure_placeholders_are_valid

    # We set the root-capability in the env, as we need it repeatedly
    @scheme_env[:"root-capability"] = @root_capability

    eval_scheme_with_args(
      "(run-suite suite_name suite_version root-capability only_tests only_capabilities exclude exclude_capabilities expected_failures)",
      suite_name: @suite_name,
      suite_version: @suite_version,
      only_tests: config.only_tests,
      only_capabilities: config.only_capabilities,
      exclude: config.exclude,
      exclude_capabilities: config.exclude_capabilities,
      expected_failures: config.expected_failures || []
    )
  end

  class Config
    attr_accessor :only_tests, :only_capabilities, :exclude, :exclude_capabilities, :expected_failures
  end
end

def suite(file_name, &block)
  config = PortsSuite::Config.new
  obj = PortsSuite.new(file_name)
  obj.instance_exec(config, &block)
  obj.run(config)
end

File: ports-js/ports.js
import { readFileSync } from 'fs';
import { Sym, Env, globalEnv, Procedure, evalSchemeString, asString } from './scheme.js';
import path from 'path';

class PortsFunction {
    constructor(func, env) {
        this.function = func;
        this.env = env;
    }
    
    isValid() {
        return this.function != null;
    }

    call(that, ...args) {
        return this.function(this.env, ...args);
    }
    
    get(key) {
        if (key === 0) {
            return Sym(this.portsRole());
        } else if (key === 1) {
            return this;
        } else {
            return null;
        }
    }
}

class PortsSetup extends PortsFunction {
    portsRole() {
        return "setup";
    }
}

class PortsTearDown extends PortsFunction {
    portsRole() {
        return "tearDown";
    }
}

class Placeholder extends PortsFunction {
    constructor(name, parameters, docString) {
        super(null, null);
        this.name = name;
        this.parameters = parameters;
        this.docString = docString;
    }
    
    get(key) {
        if (key === 0) {
            return Sym("placeholder");
        } else {
            return null;
        }
    }
}

class PortsAssertionError extends Error {
    constructor(message) {
        super(message);
        this.name = "PortsAssertionError";
    }
}

function portsAssert(value, msg = "") {
    if (!value) throw new PortsAssertionError(msg || "Assertion failed");
}

function portsAssertEq(expected, actual) {
    if (expected !== actual) throw new PortsAssertionError(`${expected} != ${actual}`);
}

class PortsSuite {
    constructor(fileName) {
        this.schemeEnv = new Env([],[], globalEnv);
        this.initializePortsPrimitives();
        this.initializePorts();
        this.suiteSource = readFileSync(fileName, 'utf8');
        
        this.suiteName = null;
        this.suiteVersion = null;
        this.sources = null;
        this.placeholders = null;
        this.rootCapability = null;
        
        this.placeholderFunctions = {};
        this.setUpFunctions = [];
        this.tearDownFunctions = [];
    }

    initializeSuite() {
        [this.suiteName, this.suiteVersion, this.sources, 
         this.placeholders, this.rootCapability] = this.evalScheme(this.suiteSource);
    }

    evalScheme(code, env = null) {
        if (!env) {
            env = this.schemeEnv;
        }
        return evalSchemeString(code, env);
    }

    evalSchemeWithArgs(code, kwargs) {
        const env = new Env([], [], this.schemeEnv);
        Object.entries(kwargs).forEach(([key, value]) => {
            env.set(Sym(key), value);
        });
        return this.evalScheme(code, env);
    }

    createPlaceholder(name, parameters, docString = "") {
        const newPlaceholder = new Placeholder(name, parameters, docString);
        this.schemeEnv.set(name, newPlaceholder);
        newPlaceholder.env = this.schemeEnv;
        
        if (this.placeholderFunctions[name]) {
            newPlaceholder.function = this.placeholderFunctions[name];
        }
        
        return newPlaceholder;
    }

    initializePortsPrimitives() {
        const primitives = {
            "create-placeholder": (...args) => this.createPlaceholder(...args),
            "is-placeholder?": x => x instanceof Placeholder,
            "assert": portsAssert,
            "assert-equal": portsAssertEq,
            "is-assertion-error?": e => {return e instanceof PortsAssertionError},
       };
        
        Object.keys(primitives).forEach((key) => {
            this.schemeEnv.set(Sym(key), primitives[key]);
        });
    }

    initializePorts() {
        const portsContent = readFileSync("ports/ports.scm", "utf8");
        this.evalScheme(portsContent);
    }

    placeholder(name, func) {
        this.placeholderFunctions[Sym(name)] = func;
        return func;
    }

    setUp(func) {
        this.setUpFunctions.push(func);
        return func;
    }

    tearDown(func) {
        this.tearDownFunctions.push(func);
        return func;
    }

    ensurePlaceholdersAreValid() {
        const invalidPlaceholders = this.placeholders.filter(p => !p.isValid());
        if (invalidPlaceholders.length > 0) {
            const invalidPlaceholderList = invalidPlaceholders.map(p => `- ${asString(p.name)}`).join("\n");
            const invalidPlaceholdersSuggestion = invalidPlaceholders.map(
                p => `suite.placeholder("${asString(p.name)}",\n\tfunction ${asString(p.name).replace(/-/g, '_')}(env, ...args) {\n\t// Implementation needed\n});`
            ).join("\n\n");
            
            throw new Error(`Empty placeholders:\n${invalidPlaceholderList}\n\nFix based on:\n${invalidPlaceholdersSuggestion}`);
        }
    }
inval
    installSetUpTearDownFunctions() {
        this.setUpFunctions.forEach(func => {
            this.rootCapability[2].unshift(new PortsSetup(func, this.schemeEnv));
        });
        this.tearDownFunctions.forEach(func => {
            this.rootCapability[3].unshift(new PortsTearDown(func, this.schemeEnv));
        });
    }

    run({only = null, onlyCapabilities = null, exclude = null, excludeCapabilities = null, expectedFailures = []} = {}) {
        this.initializeSuite();
        this.installSetUpTearDownFunctions();
        this.ensurePlaceholdersAreValid();
        
        // We set the root-capability in the env, as we need it repeatedly
        this.schemeEnv.set(Sym("root-capability"), this.rootCapability);

        this.evalSchemeWithArgs(
            "(run-suite suite_name suite_version root-capability only_tests only_capabilities exclude exclude_capabilities expected_failures)", 
            {suite_name: this.suiteName,
            suite_version: this.suiteVersion,
            only_tests: only, 
            only_capabilities: onlyCapabilities, 
            exclude: exclude, 
            exclude_capabilities: excludeCapabilities,
            expected_failures: expectedFailures});
    }
}

export function suite(fileName) {
    return new PortsSuite(fileName);
}

